### PR TITLE
backend: (riscv) add optimisations for memref store/load

### PR DESCRIPTION
**Pull request by superlopuh**
This adds just the necessary optimisation patterns to ensure that we can constant fold our way to a single instruction when the offsets are constants.

- [x] add tests for each individual optimisation
- [x] add integer store/load

### CODE REVIEW

**Review by nazavode — APPROVED** [review]
Super neat!

File: tests/filecheck/backend/riscv/memref_to_riscv_opt.mlir [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+builtin.module {
+    %v_reg, %m_reg = "test.op"() : () -> (!riscv.freg<>, !riscv.reg<>)
+    %v = builtin.unrealized_conversion_cast %v_reg : !riscv.freg<> to f32
+    %m = builtin.unrealized_conversion_cast %m_reg : !riscv.reg<> to memref<3x2xf32>
+    %r_reg = riscv.li 1 : () -> !riscv.reg<>
+    %c_reg = riscv.li 1 : () -> !riscv.reg<>
+    %r = builtin.unrealized_conversion_cast %r_reg : !riscv.reg<> to index
+    %c = builtin.unrealized_conversion_cast %c_reg : !riscv.reg<> to index
+    "memref.store"(%v, %m, %r, %c) {"nontemporal" = false} : (f32, memref<3x2xf32>, index, index) -> ()
+    %x = "memref.load"(%m, %r, %c) {"nontemporal" = false} : (memref<3x2xf32>, index, index) -> (f32)
```
Comment (JosseVanDelm) [inline tests/filecheck/backend/riscv/memref_to_riscv_opt.mlir:14]: Does this also work for ints?

**Review by JosseVanDelm — COMMENTED** [review]
Hi, this looks really cool!
Maybe not related to this PR,
Do we have support for ints as well?

File: tests/filecheck/backend/riscv/memref_to_riscv_opt.mlir (hunk 1 above)
Comment (adutilleul) [reply tests/filecheck/backend/riscv/memref_to_riscv_opt.mlir:14]: The lowering from memref also handles integer types at least (https://github.com/xdslproject/xdsl/blob/8f06da92e84347a59764cc4fbc462fa6c1a37b95/docs/Toy/toy/rewrites/lower_memref_riscv.py#L108).
It should be also quite straightforward to add the optimization for `lw/sw'.

Comment (JosseVanDelm) [reply tests/filecheck/backend/riscv/memref_to_riscv_opt.mlir:14]: I added a comment for this on #1507

Comment (superlopuh) [reply tests/filecheck/backend/riscv/memref_to_riscv_opt.mlir:14]: Definitely related to this PR, but I'd rather add it in a follow-up if that's ok with you?

**Review by adutilleul — APPROVED** [review]

**Pull request merged**